src/compas/datastructures/mesh/join.py

```diff
@@ -4,57 +4,105 @@
     'meshes_join',
+    'meshes_join_and_weld'
 ]
 
+def meshes_join(meshes, cls = None):
+    """Join meshes without welding.
+    Parameters
+    ----------
+    meshes : list
+        A list of meshes.
+    Returns
+    -------
+    mesh
+        The joined mesh.
+    """
+
+    if cls is None:
+        cls = type(meshes[0])
 
-def meshes_join(meshes, delete_duplicates=False, precision='3f'):
-    """Join multiple meshes. These meshes are usually assumed
-    to be mesh patches that nicely align along their boundaries.
+    vertices = []
+    faces = []
+
+    for mesh in meshes:
+        # create vertex map based on geometric keys in dictionary with duplicates
+        vertex_map = ({vkey: len(vertices) + i for i, vkey in enumerate(mesh.vertices())})
+        # list vertices with coordinates
+        vertices += [mesh.vertex_coordinates(vkey) for vkey in mesh.vertices()]
+        # modify vertex indices in the faces
+        faces += [ [vertex_map[vkey] for vkey in mesh.face_vertices(fkey)] for fkey in mesh.faces()]
 
+    return cls.from_vertices_and_faces(vertices, faces)
+
+def mesh_weld(mesh, precision = None, cls = None):
+    """Weld vertices of a mesh within some precision distance.
     Parameters
     ----------
-    meshes : list of Mesh
-        A list of mesh objects.
-    delete_duplicates: bool
-        True, if resulting duplicate vertices should be deleted.
-        False, otherwise.
-
+    mesh : Mesh
+        A mesh.
+    precision: str
+        Tolerance distance for welding.
+    Returns
+    -------
+    mesh
+        The welded mesh.
     """
-    count = 0
-    mesh_all = deepcopy(meshes[0])
-    mesh_all.clear()
-    key_map = {}
-    for mesh in meshes:
-        faces = list(mesh.faces())
-        vertices = list(mesh.faces())
 
-        for key, attr in mesh.vertices(True):
-            mesh_all.add_vertex(count, x=attr['x'], y=attr['y'], z=attr['z'], attr_dict=attr)
-            key_map[key] = count
-            count += 1
+    if cls is None:
+        cls = type(mesh)
+
+    # create vertex map based on geometric keys in dictionary without duplicates
+    vertex_map = {geometric_key(mesh.vertex_coordinates(vkey), precision): vkey for vkey in mesh.vertices()}
+    # list vertices with coordinates
+    vertices = [reverse_geometric_key(geom_key) for geom_key in vertex_map.keys()]
+    # reorder vertex keys in vertex map
+    vertex_map = {geom_key: i for i, geom_key in enumerate(vertex_map.keys())}
+    # modify vertex indices in the faces
+    faces = [ [vertex_map[geometric_key(mesh.vertex_coordinates(vkey), precision)] for vkey in mesh.face_vertices(fkey)] for fkey in mesh.faces()]
 
-        for fkey,attr in mesh.faces(True):
-            vertices = mesh.face_vertices(fkey)
-            new_vertices = [key_map[key] for key in vertices]
-            mesh_all.add_face(new_vertices)
+    return cls.from_vertices_and_faces(vertices, faces)
+
+def meshes_join_and_weld(meshes, precision = None, cls = None):
```

Comment:
no spaces around `=` in assignment of default values for arguments

Reply:
also should `weld` not be an optional flag in `meshes_join` instead of a separate function?